### PR TITLE
Bug 1727024: Add OperatorUpgradeable status to StatusReporter

### PR DESCRIPTION
**Pull request by s-urbaniak**
Currently we don't report that status. This fixes it.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1727024

/cc @LiliC @paulfantom

### CODE REVIEW

**Comment by s-urbaniak** [comment]
/hold

let me add some more tests for the status reporter.

File: pkg/client/status_reporter.go [hunk 1]
```diff
@@ -89,6 +90,7 @@ func (r *StatusReporter) SetInProgress() error {
 
 	conditions := newConditions(co.Status, r.version, time)
 	conditions.setCondition(v1.OperatorProgressing, v1.ConditionTrue, "Rolling out the stack.", time)
+	conditions.setCondition(v1.OperatorUpgradeable, v1.ConditionFalse, "", time)
```
Comment (LiliC) [inline pkg/client/status_reporter.go]: In the API docs for `OperatorUpgradeable`, it says we should set a reason for when this is false:

> and the message field should contain a human readable description of what the administrator should do to allow the operator to successfully update.

File: pkg/client/status_reporter.go [hunk 2]
```diff
@@ -112,6 +114,7 @@ func (r *StatusReporter) SetFailed(statusErr error) error {
 	conditions.setCondition(v1.OperatorAvailable, v1.ConditionFalse, "", time)
 	conditions.setCondition(v1.OperatorProgressing, v1.ConditionFalse, "", time)
 	conditions.setCondition(v1.OperatorDegraded, v1.ConditionTrue, fmt.Sprintf("Failed to rollout the stack. Error: %v", statusErr), time)
+	conditions.setCondition(v1.OperatorUpgradeable, v1.ConditionFalse, "", time)
```
Comment (LiliC) [inline pkg/client/status_reporter.go]: Same as above, we should set the msg here to why this is not upgradable currently.

File: pkg/client/status_reporter.go (hunk 1 above)
Comment (s-urbaniak) [reply pkg/client/status_reporter.go]: great catch :+1: fixed

File: pkg/client/status_reporter.go (hunk 2 above)
Comment (s-urbaniak) [reply pkg/client/status_reporter.go]: fixed

**Comment by s-urbaniak** [comment]
@LiliC addressed comments, PTAL

File: pkg/client/condition.go [hunk 3]
```diff
@@ -41,6 +41,11 @@ func newConditions(cos v1.ClusterOperatorStatus, targetVersion string, time meta
 			Status:             v1.ConditionUnknown,
 			LastTransitionTime: time,
 		},
+		v1.OperatorUpgradeable: {
+			Type:               v1.OperatorUpgradeable,
+			Status:             v1.ConditionUnknown,
```
Comment (LiliC) [inline pkg/client/condition.go]: So we are fine with setting this to default as `ConditionUnknown` as this means more or less that it's possible to upgrade?

File: pkg/client/status_reporter.go [hunk 4]
```diff
@@ -89,6 +90,10 @@ func (r *StatusReporter) SetInProgress() error {
 
 	conditions := newConditions(co.Status, r.version, time)
 	conditions.setCondition(v1.OperatorProgressing, v1.ConditionTrue, "Rolling out the stack.", time)
+	conditions.setCondition(v1.OperatorUpgradeable, v1.ConditionFalse,
+		"Rollout of the monitoring stack is progressing. Please wait until it finishes.",
```
Comment (LiliC) [inline pkg/client/status_reporter.go]: Nit:
```suggestion
		"Rollout of the monitoring stack is in progress. Please wait until it finishes.",
```

File: pkg/client/condition.go (hunk 3 above)
Comment (s-urbaniak) [reply pkg/client/condition.go]: This is just the constructor for the transient set of states in this state machine. The actually persisted states are as follows (as is asserted in the tests):

1. "In Progress" (aka initally persisted state):
"Available": "Unknown",
"Degraded": "Unknown",
"Progressing": "True",
**"Upgradeable", "False"** <- this is debatable, we could also leave this as "Unknown", but we probably don't want to upgrade during rollout.

2. "Done" (aka happy case):
"Available", "True",
"Degraded", "False",
"Progressing", "False",
**"Upgradeable", "True"**

3. "Failed" (aka unhappy case):
"Available", "False",
"Degraded", "True",
"Progressing", "False",
**"Upgradeable", "False"**

File: pkg/client/status_reporter.go (hunk 4 above)
Comment (s-urbaniak) [reply pkg/client/status_reporter.go]: oh good catch :+1:

**Pull request merged**